In [31]:
from sklearn import datasets
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
import pandas as pd

In [32]:
# load diabetes datasets
X, y = datasets.load_diabetes(as_frame=True, scaled=False, return_X_y=True)

### Perform cross-validation on nine polynomial models, ranging from degree 0 to 8.

In [33]:
degrees = range(9)
cross_val = []
r2_score = []
for degree in degrees:
    model = make_pipeline(PolynomialFeatures(degree),LinearRegression())
    scores = cross_val_score(model,X,y,cv=5,scoring='neg_mean_absolute_error').mean()
    r2_scores = cross_val_score(model,X,y,cv=5,scoring='r2').mean()
    cross_val.append(-scores)
    r2_score.append(r2_scores)

### Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values.

In [34]:
#Cross-validation result summary
Cross_Summary = pd.DataFrame({"Degree":degrees, "r2_Score":r2_score, "Cross_Val_Score(MAE)":cross_val})
Cross_Summary

Degree      r2_Score  Cross_Val_Score(MAE)
0       0     -0.027506             66.045624
1       1      0.482316             44.276499
2       2     -1.485182             80.595439
3       3   -203.418938            342.051924
4       4   -571.083108            657.260477
5       5   -436.856887            562.993636
6       6  -1694.550210            742.717488
7       7  -5530.894075           1032.681725
8       8 -16076.255118           1475.658532

In [35]:
#Calculate mean and standard deviation
Mean_r2 = Cross_Summary['r2_Score'].mean()
SD_r2 = Cross_Summary['r2_Score'].std()
Mean_MAE = Cross_Summary['Cross_Val_Score(MAE)'].mean()
SD_MAE = Cross_Summary['Cross_Val_Score(MAE)'].std()

In [36]:
print("\nMean R-Squared is", Mean_r2)
print("Standard Deviation R-Squared is", SD_r2)
print("Mean MAE:", Mean_MAE)
print("Standard Deviation MAE:", SD_MAE)


Mean R-Squared is -2723.787633985016
Standard Deviation R-Squared is 5311.490588389923
Mean MAE: 556.0312604351241
Standard Deviation MAE: 486.29502330395763


### Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model.

The best-performing model is the one with the highest R-squared and the lowest MAE. A higher R-squared indicates a better fit to the data, as more variance is accounted for. A lower MAE indicates better accuracy and a smaller average error in predictions.

According to above data model with degree of 1 has a highest R-squared value (0.482316) and lowest Mean Absolute Error (44.276499). Therefore, model with the degree of 1 exhibits the highest performance.